exec/java-exec/src/main/java/org/apache/drill/exec/planner/physical/visitor/SplitUpComplexExpressions.java

```diff
@@ -44,103 +44,108 @@
 
-    for (RexNode rex : project.getChildExps()) {
-      origRelDataTypes.add(project.getRowType().getFieldList().get(i));
-      i++;
-      exprList.add(rex.accept(exprSplitter));
+    List<RelDataTypeField> projectFields = newProject.getRowType().getFieldList();
+    List<RelDataTypeField> origRelDataTypes = new ArrayList<>();
+    List<RexNode> exprList = new ArrayList<>();
+    final int lastRexInput = lastColumnReferenced + 1;
+    RexVisitorComplexExprSplitter exprSplitter = new RexVisitorComplexExprSplitter(funcReg, rexBuilder, lastRexInput);
+    int i = 0;
+    for (RexNode rex : newProject.getChildExps()) {
+      RelDataTypeField originField = projectFields.get(i++);
+      RexNode splitRex = rex.accept(exprSplitter);
+      origRelDataTypes.add(originField);
+      exprList.add(splitRex);
     }
-    List<RexNode> complexExprs = exprSplitter.getComplexExprs();
 
-    if (complexExprs.size() == 1 && findTopComplexFunc(project.getChildExps()).size() == 1) {
-      return project;
+    final List<RexNode> complexExprs = exprSplitter.getComplexExprs();
+    if (complexExprs.size() == 1 && findTopComplexFunc(newProject.getChildExps()).size() == 1) {
+      return newProject;
     }
 
-    ProjectPrel childProject;
 
-    List<RexNode> allExprs = new ArrayList<>();
-    int exprIndex = 0;
-    List<String> fieldNames = originalInput.getRowType().getFieldNames();
-    for (int index = 0; index < lastRexInput; index++) {
-      RexBuilder builder = new RexBuilder(factory);
-      allExprs.add(builder.makeInputRef( new RelDataTypeDrillImpl(new RelDataTypeHolder(), factory), index));
 
-      if (fieldNames.get(index).contains(SchemaPath.DYNAMIC_STAR)) {
-        relDataTypes.add(new RelDataTypeFieldImpl(fieldNames.get(index), allExprs.size(), factory.createSqlType(SqlTypeName.ANY)));
-      } else {
-        relDataTypes.add(new RelDataTypeFieldImpl("EXPR$" + exprIndex, allExprs.size(), factory.createSqlType(SqlTypeName.ANY)));
-        exprIndex++;
-      }
-    }
-    RexNode currRexNode;
-    int index = lastRexInput - 1;
 
     // if the projection expressions contained complex outputs, split them into their own individual projects
     if (complexExprs.size() > 0 ) {
+      List<RexNode> allExprs = new ArrayList<>();
+      int exprIndex = 0;
+      List<String> fieldNames = newInput.getRowType().getFieldNames();
+
+      List<RelDataTypeField> relDataTypes = new ArrayList<>();
+      for (int index = 0; index < lastRexInput; index++) {
+        allExprs.add(rexBuilder.makeInputRef( new RelDataTypeDrillImpl(new RelDataTypeHolder(), factory), index));
+
+        if (fieldNames.get(index).contains(SchemaPath.DYNAMIC_STAR)) {
+          relDataTypes.add(new RelDataTypeFieldImpl(fieldNames.get(index), allExprs.size(), factory.createSqlType(SqlTypeName.ANY)));
+        } else {
+          relDataTypes.add(new RelDataTypeFieldImpl("EXPR$" + exprIndex, allExprs.size(), factory.createSqlType(SqlTypeName.ANY)));
```

Comment:
Please use `SqlValidatorUtil.EXPR_SUGGESTER`:
```suggestion
          relDataTypes.add(new RelDataTypeFieldImpl(SqlValidatorUtil.EXPR_SUGGESTER.apply(null, exprIndex, 0), allExprs.size(), factory.createSqlType(SqlTypeName.ANY)));
```